cmake/extensions.cmake

```diff
@@ -1711,56 +1711,201 @@ endfunction()
 to absolute path using `\${CMAKE_CURRENT_SOURCE_DIR}/${path}` or similar. \n \
 Relative paths are only allowed with `-D${ARGV1}=<path>`")
       endif()
     endforeach()
 
     # This updates the provided argument in parent scope (callers scope)
-    set(${ARGV1} ${${ARGV1}} PARENT_SCOPE)
+    set(${FILE_APPLICATION_ROOT} ${${FILE_APPLICATION_ROOT}} PARENT_SCOPE)
+  endif()
+
+  if(FILE_CONF_FILES)
+    set(FILENAMES ${BOARD})
+
+    if(DEFINED BOARD_REVISION)
+      list(APPEND FILENAMES "${BOARD}_${BOARD_REVISION_STRING}")
+    endif()
+
+    if(FILE_DTS)
+      foreach(filename ${FILENAMES})
+        if(EXISTS ${FILE_CONF_FILES}/${filename}.overlay)
+          list(APPEND ${FILE_DTS} ${FILE_CONF_FILES}/${filename}.overlay)
+        endif()
+      endforeach()
+
+      # This updates the provided list in parent scope (callers scope)
+      set(${FILE_DTS} ${${FILE_DTS}} PARENT_SCOPE)
+    endif()
+
+    if(FILE_KCONF)
+      foreach(filename ${FILENAMES})
+        if(FILE_BUILD)
+          set(filename "${filename}_${FILE_BUILD}")
+        endif()
+
+        if(EXISTS ${FILE_CONF_FILES}/${filename}.conf)
+          list(APPEND ${FILE_KCONF} ${FILE_CONF_FILES}/${filename}.conf)
+        endif()
+      endforeach()
+
+      # This updates the provided list in parent scope (callers scope)
+      set(${FILE_KCONF} ${${FILE_KCONF}} PARENT_SCOPE)
+    endif()
   endif()
 endfunction()
 
+# Usage:
+#   zephyr_check_cache(<variable> [REQUIRED])
+#
+# Check the current CMake cache for <variable> and warn the user if the value
+# is being modified.
+#
+# This can be used to ensure the user does not accidentally try to change
+# Zephyr build variables, such as:
+# - BOARD
+# - SHIELD
+#
+# variable: Name of <variable> to check and set, for example BOARD.
+# REQUIRED: Optional flag. If specified, then an unset <variable> will be
+#           treated as an error.
+#
+# Details:
+#   <variable> can be set by 3 sources. Through environment variables,
+#   through the cmake CLI, and through CMakeLists.txt.
+#
+#   CLI has the highest precedence, then comes environment variables,
+#   and then finally CMakeLists.txt.
+#
+#   A user can ignore all the precedence rules if the same source is always used
+#   E.g. always specifies -D<variable>= on the command line,
+#   always has an environment <variable> set, or always has a set(<variable> foo)
+#   line in his CMakeLists.txt and avoids mixing sources.
+#
+#   The selected <variable> can be accessed through the variable '<variable>' in
+#   following Zephyr CMake code.
+#
+#   If the user tries to change <variable> to a new value, then a warning will
+#   be printed, and the previously value (CACHED_<variable>) will be used, as it
+#   has precedence.input sources, and the
+#   previously used.
+#   Together with the warning, user is informed that in order to change <variable>
+#   the board directory must be cleaned.
+#
+function(zephyr_check_cache variable)
```

Comment:
Some things about this function:

1. It still has a copy/pasted `BOARD` in a comment
2. It's changing behavior for shields. Only BOARD_DEPRECATED exists. SHIELD_DEPRECATED does not exist. This function shouldn't be involved in the DEPRECATED variable checks; that should be the caller's responsibility in the case of the BOARD variable. This function should not check `${variable}_DEPRECATED` at all.
3. I understand the idea behind `variable_text`, but I think the error message would be much clearer if the actual variable was printed unaltered (and for it to get upgraded to a fatal error instead of a warning, but maybe that's too much for one commit)
4. The CMake documentation seems to discourage the `endfunction(name ...)` form, saying it "is supported for backward compatibility only", so why continue to use it in new code?
5. There's no need for fancy tricks like `${variable}_cli_argument`. This is a function and has its own scope, so we don't need to worry about polluting it.
6. I think the comment about "there are actually 4 sources" is really useful for understanding and should not be removed.

Combined, I am suggesting something like this (not tested):

```cmake
function(zephyr_check_cache variable)
  cmake_parse_arguments(CACHE_VAR "REQUIRED" "" "" ${ARGN})
  string(TOLOWER ${variable} variable_text)
  string(REPLACE "_" " " variable_text ${variable_text})

  get_property(cached_value CACHE ${variable} PROPERTY VALUE)

  # There are actually 4 sources: the three user input sources, and
  # any previously used value we stored in CACHED_${variable}.
  #
  # The previously used value has precedence, and is how we detect
  # that the user is trying to change the value in order to warn about
  # needing to do a pristine build.

  set(cur_value ${cached_value}) # Either new or old
  if(cur_value STREQUAL CACHED_${variable})
    # CACHED_${variable} agrees with cur_value, so no need to keep it.
    unset(cur_value)
  endif()

  set(app_cmake_lists ${${variable}})
  if(cached_value STREQUAL ${variable})
    # The app build scripts did not set a default, The BOARD we are
    # reading is the cached value from the CLI
    unset(app_cmake_lists)
  endif()

  if(DEFINED CACHED_${variable})
    # Warn the user if it looks like he is trying to change the board
    # without cleaning first
    if(cur_value)
      if(NOT ((CACHED_${variable} STREQUAL cur_value)))
        message(FATAL_ERROR "${variable} was previously set to ${CACHED_${variable}}, and cannot be changed to ${cur_value}. To fix, do a pristine build.")
      endif()
    endif()

    if(CACHED_${variable})
      set(${variable} ${CACHED_${variable}} PARENT_SCOPE)
      # This resets the user provided value with previous (working) value.
      set(${variable} ${CACHED_${variable}} CACHE STRING "Selected ${variable}" FORCE)
    else()
      unset(${variable} PARENT_SCOPE)
      unset(${variable} CACHE)
    endif()
  elseif(cur_value)
    set(${variable} ${cur_value} PARENT_SCOPE)

  elseif(DEFINED ENV{${variable}})
    set(${variable} $ENV{${variable}} PARENT_SCOPE)

  elseif(app_cmake_lists)
    set(${variable} ${app_cmake_lists} PARENT_SCOPE)

  elseif(${CACHE_VAR_REQUIRED})
    message(FATAL_ERROR "${variable} was not set on the CMake command-line, in the environment, or by the application CMakeLists.txt.")
  endif()

  # Store the specified variable in the cache
  set(CACHED_${variable} ${${variable}} CACHE STRING "Selected ${variable}")
endfunction()
```

Reply:
> Some things about this function:
> 

Thanks for the thorough review.
It's always good to take a critical view on existing code, even when it's just being moved around.

>     1. It still has a copy/pasted `BOARD` in a comment

Will fix.

>     2. It's changing behavior for shields. Only BOARD_DEPRECATED exists. SHIELD_DEPRECATED does not exist. This function shouldn't be involved in the DEPRECATED variable checks; that should be the caller's responsibility in the case of the BOARD variable. This function should not check `${variable}_DEPRECATED` at all.

Correct that SHIELD_DEPRECATED is not in use today, but allowing for `${variable}_DEPRECATED` will ensure that we keep a uniform way for deprecating Zephyr CMake argument values that are sticky.
I prefer to keep it, as that will allow shield to use this functionality in future and minimize the risk that things are done slightly different. (and will allow for other sticky settings to use deprecated feature)
Just as example, in samples, Kconfig uses `prj.conf` but dts uses `app.overlay`, but for boards they both uses `<board>.conf` or `<board>.overlay`. If we have had some common code in place from beginning, then those name might have been `prj.conf` and `prj.overlay`.

Just because shield today doesn't deprecate names, that doesn't mean they won't do so in future.
And the same could be true for other Zephyr CMake settings introduced in future.

>     3. I understand the idea behind `variable_text`, but I think the error message would be much clearer if the actual variable was printed unaltered
> 

In certain cases it could be the variable is not set, but agree, printing the values would be nice improvement.

>  (and for it to get upgraded to a fatal error instead of a warning, but maybe that's too much for one commit)

That would be a change of behavior to current handling that I think should be discussed separately.

>     4. The CMake documentation seems to discourage the `endfunction(name ...)` form, saying it "is supported for backward compatibility only", so why continue to use it in new code?
 
Correct. 
I don't use this form so often, but with longer functions and in a big file with many functions (like extensions.cmake), I feel it helps to keep track of which function is actually ending. See it similar to an `#endif // SOMETHING`.

I would find it weird in CMake to write `endfunction() # zephyr_check_cache` now that CMake does support the `endfunction(zephyr_check_cache)` variant.

and the form is not deprecated.

>     5. There's no need for fancy tricks like `${variable}_cli_argument`. This is a function and has its own scope, so we don't need to worry about polluting it.
 
The name `cli_argument` was not due to collision, but `-D<variable>=<newvalue>` goes direct into the cache, so what we are testing is actually if user has provided a new `-D<variable>` value, so the name `cli_argument` is well chosen, and was already used in previous code:
https://github.com/zephyrproject-rtos/zephyr/blob/4d930057d6d667634752e3d0d0dbd677761714e2/cmake/app/boilerplate.cmake#L189
https://github.com/zephyrproject-rtos/zephyr/blob/4d930057d6d667634752e3d0d0dbd677761714e2/cmake/app/boilerplate.cmake#L270
but the `${variable}` part is not needed anymore, I agree to that.

>     6. I think the comment about "there are actually 4 sources" is really useful for understanding and should not be removed.
> 

I will try to rephrase then, cause I found the existing 3 sources and then 4 sources to be a bit confusing.
Probably be more explicit on 3 sources for setting the var, and then having a cached value to ensure that user doesn't try to do something unsupported.

Reply:
> >   1. It still has a copy/pasted `BOARD` in a comment
>
> Will fix.

Fixed.



> > ```
> > 3. I understand the idea behind `variable_text`, but I think the error message would be much clearer if the actual variable was printed unaltered
> > ```
> 
> 
> In certain cases it could be the variable is not set, but agree, printing the values would be nice improvement.

Improved, now printing current value, and ignored value.



> > ```
> > 5. There's no need for fancy tricks like `${variable}_cli_argument`. This is a function and has its own scope, so we don't need to worry about polluting it.
> > ```
 
`${variable}_` removed.



> > ```
> > 6. I think the comment about "there are actually 4 sources" is really useful for understanding and should not be removed.
> > ```
> 
> 
> I will try to rephrase

Added additional inline comment and extended and cleaned up the function description.